docs: add troubleshooting guide for service mesh (#46633)
sourcegraph/sourcegraph: doc/admin/deploy/kubernetes/troubleshoot.md

    port: minio   #this port name MUST exist in the same spec
```

## Service mesh

Known issues when using a service mesh (e.g. Istio, Linkerd, etc.)

#### Error message: `Git command [git rev-parse HEAD] failed (stderr: ""): strconv.Atoi: parsing "": invalid syntax`

<img class="screenshot w-100" src="https://user-images.githubusercontent.com/68532117/178506378-3d047bc5-d672-487a-920f-8f228ae5cb27.png"/>

This error occurs because Envoy, the proxy used by Istio, [drops proxied trailers for the requests made over HTTP/1.1 protocol by default](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/protocol.proto#config-core-v3-http1protocoloptions). To resolve this issue, enable trailers in your instance following the examples provided for [Kubernetes](https://github.com/sourcegraph/deploy-sourcegraph/tree/master/overlays) and [Kubernetes with Helm](https://github.com/sourcegraph/deploy-sourcegraph-helm/tree/main/charts/sourcegraph/examples).

### Symbols sidebar and hovers are not working

<img class="screenshot w-100" src="https://user-images.githubusercontent.com/68532117/212372086-4c53e817-be3d-46b6-9cc1-fc34e695d30c.png"/>

> This issue occurs when the "symbols" component attempts to connect to other services using the built-in service discovery system provided by Sourcegraph, instead of using the service discovery system provided by Kubernetes or a service mesh (such as Istio).

Reviewer: This is incorrect. We are using kubernetes service discovery. Our code watches the `symbols` endpoint. IE it is kind of like the command `kubectl get endpoints symbols --watch`. From the endpoint resources we then connect directly to the pods IP address.

> This issue occurs when the "symbols" component attempts to connect to other services using the built-in service discovery system provided by Sourcegraph, instead of using the service discovery system provided by Kubernetes or a service mesh (such as Istio).

Author: Replace with an explanation that focuses on requests being intercepted due to IP addresses.

In general, service meshes like Istio use a feature called mutual Transport Layer Security (mTLS) to secure communication between services. mTLS relies on services communicating with each other using DNS names, rather than IP addresses. These DNS names are used to identify the specific services or pods that the communication is intended for.

> In this case, when the Envoy sidecar (a component used to manage communication between services) intercepts a request from the "frontend" component to the "symbols" component, it is unable to locate the correct upstream service because it is using the IP address instead of the DNS name.

Reviewer: This is very enlightening thanks. In fact it makes me realise we can probably fix this in our code. In some cases we can construct the DNS name of a pod instead of using the IP. I would be interested to know if we can just use the pod name, or if we need a more fully qualified name? Additionally it would be interesting to know if we could switch to using pod DNS names by default, or if it should be opt in (I worry about breaking environment were pod dns names do not resolve).

But yeah, check this example output from our cluster, you can see it in includes pod names in the endpoint resource:

``` shellsession
❯ kubectl get -o yaml endpoints symbols
...snip
subsets:
- addresses:
  - ip: 10.164.14.148
    nodeName: gke-cloud-containerd-pool-3ff685fe-b4ks
    targetRef:
      kind: Pod
      name: symbols-85fdd66bc6-p9rxp
      namespace: prod
      resourceVersion: "753143567"
      uid: 5661f7e1-c2a5-4488-892f-22a5e14dd7a2
snip...
```

For example, in a good request, the communication flow would be: frontend -> http://symbol:3184/ -> envoy -> [look up upstream service using DNS name] -> envoy -> symbols

> In a bad request, the communication flow would be: frontend -> http://symbol_pod_ip:3184 -> envoy -> [symbol_pod_ip not found since it's an IP address not a DNS name] -> envoy -> symbols

Reviewer 2: I remember we have a sample payload when this error occurs, maybe put it here? (502/503 upstream not found or something)

> In a bad request, the communication flow would be: frontend -> http://symbol_pod_ip:3184 -> envoy -> [symbol_pod_ip not found since it's an IP address not a DNS name] -> envoy -> symbols

Author: @michaellzc I think `502/503 upstream not found or something` was caused by another setting that the customer had. When I tried to reproduce it locally, the 502 error did not show up. Instead, it was just waiting for response indefinitely

> To resolve this issue, the solution is to redeploy the frontend after specifying the service address for symbols by setting the SYMBOLS_URL environment variable in frontend.

Reviewer: This will make having more than one replica of symbols ineffective. IE we will not shard requests based on repository/commit anymore. IE I would call out you can only use one pod in this case.

> To resolve this issue, the solution is to redeploy the frontend after specifying the service address for symbols by setting the SYMBOLS_URL environment variable in frontend.

Author: Added a WARNING: This option is recommended only for symbols with a single replica. Enabling this option will negatively impact the performance of the symbols service when it has multiple replicas, as it will no longer be able to distribute requests by repository/commit.

    port: minio   #this port name MUST exist in the same spec
```

## Service mesh

> Known issues related to service mesh. E.g. Istio, Linkerd, etc. 

Reviewer 3: ```suggestion
Known issues when using a service mesh (e.g. Istio, Linkerd, etc.)
```

#### Error message: `Git command [git rev-parse HEAD] failed (stderr: ""): strconv.Atoi: parsing "": invalid syntax`

<img class="screenshot w-100" src="https://user-images.githubusercontent.com/68532117/178506378-3d047bc5-d672-487a-920f-8f228ae5cb27.png"/>

> This error occurs when the service mesh, like istio, drops the [Trailer response header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Trailer) for http1 by default. Enable trailers in your instance to resolve this issue. 

Reviewer 3: ```suggestion
This error occurs when a service mesh, like istio, drops the [Trailer response header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Trailer) for http1 by default. Enable trailers in your instance to resolve this issue. 
```

Reviewer 3: I'm not sure what "http1" refers to here? Is it an Istio concept? Or does it refer to HTTP/1.1 or any HTTP/1 protocol (in which case, I'd use "HTTP/1.x")

> This error occurs when a service mesh, like Istio, drops the [Trailer response header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Trailer) for http1 by default. Enable trailers in your instance to resolve this issue. 

Reviewer 3: I'm not sure what "http1" refers to here? Is it an Istio concept? Or does it refer to HTTP/1.1 or any HTTP/1 protocol (in which case, I'd use "HTTP/1.x")

> See our examples on applying the EnvoyFilter in [Kubernetes without Helm](https://github.com/sourcegraph/deploy-sourcegraph/tree/master/overlays) and [Kubernetes with Helm](https://github.com/sourcegraph/deploy-sourcegraph-helm/tree/main/charts/sourcegraph/examples) for details.

Reviewer 3: ```suggestion
See our examples on applying the EnvoyFilter in [Kubernetes](https://github.com/sourcegraph/deploy-sourcegraph/tree/master/overlays) and [Kubernetes with Helm](https://github.com/sourcegraph/deploy-sourcegraph-helm/tree/main/charts/sourcegraph/examples) for details.
```

> ### Code Intel hover or Symbols sidebar are not working

Reviewer 3: ```suggestion
### Symbols sidebar and hovers are not working
```

<img class="screenshot w-100" src="https://user-images.githubusercontent.com/68532117/212372086-4c53e817-be3d-46b6-9cc1-fc34e695d30c.png"/>

> This occurs when the symbols tries to connect using built-in service discovery by Sourcegraph instead of the service discovery provided by the Kubernetes or service mesh. In general, Istio's mTLS (as well as other popular service mesh) relies on inter-service comm being done over DNS name (it's used to identity services/pods). Therefore, when the envoy sidecar intercepts frontend's request toward symbols, it fails to locate the right upstream service (i.e. symbols) because it's using the IP address.

Reviewer 3: I found this a bit confusingly worded. Turns out we can use an LLM to translate this to clearer terms that also include some additional helpful context:
<img width="587" alt="image" src="https://user-images.githubusercontent.com/1646931/213317433-cce6bfe5-ce23-44e8-a1e1-6ff54243b872.png">

Please double-check the correctness here:

```suggestion
This issue occurs when the "symbols" component attempts to connect to other services using the built-in service discovery system provided by Sourcegraph, instead of using the service discovery system provided by Kubernetes or a service mesh (such as Istio).

In general, service meshes like Istio use a feature called mutual Transport Layer Security (mTLS) to secure communication between services. mTLS relies on services communicating with each other using DNS names, rather than IP addresses. These DNS names are used to identify the specific services or pods that the communication is intended for.

In this case, when the Envoy sidecar (a component used to manage communication between services) intercepts a request from the "frontend" component to the "symbols" component, it is unable to locate the correct upstream service because it is using the IP address instead of the DNS name.
```

For example:

> - Good request: frontend -> http://symbol:3184 -> envoy -> [look up upstream service] -> envoy -> symbols
> - Bad request:  frontend -> http://symbol_pod_ip:3184 -> envoy -> [who is symbol_pod_ip?] -> envoy -> symbols

Reviewer 3: ```suggestion
For example, in a good request, the communication flow would be: frontend -> http://symbol:3184/ -> envoy -> [look up upstream service using DNS name] -> envoy -> symbols

However, in a bad request, the communication flow would be: frontend -> http://symbol_pod_ip:3184 -> envoy -> [who is symbol_pod_ip? since it's an IP address not a DNS name] -> envoy -> symbols
```

> The solution is to redeploy frontend after specifying the service address for symbols setting the **SYMBOLS_URL** environment variables in frontend:

Reviewer 3: ```suggestion
To resolve this issue, the solution is to redeploy the frontend after specifying the service address for symbols by setting the SYMBOLS_URL environment variable in frontend.
```

```yaml
SYMBOLS_URL=http:symbols:3184
```

Please make sure the old frontend pods are removed.

> ### Squirrel.LocalCodeIntel http status 502

Reviewer 3: ```suggestion
### Squirrel.LocalCodeIntel HTTP status 502
```

<img class="screenshot w-100" src="https://user-images.githubusercontent.com/68532117/212374098-dc2dfe69-4d26-4f5e-a78b-37a53c19ef22.png"/>

> 
> Code Intel hover might be stuck in loading or return a 502 error: `Squirrel.LocalCodeIntel http status 502:`
> 
> This is due to the same issue discussed in the issue above with the same solution.

Reviewer 3: ```suggestion
The issue described is related to the Code Intel hover feature, where it may get stuck in a loading state or return a 502 error with the message `Squirrel.LocalCodeIntel http status 502`. This is caused by the same issue as above ("Symbols sidebar and hovers are not working"). See that section for the solution.
```
